File: 6-DescidaDaSerra/main.py
import sys

def main():
    tamanho_min_tanque = 0
    gasto_combustivel = []  # Lista de tuplas (distancia, consumo)
    total_vazamentos = 0
    consumo = -1  # Consumo atual

    for line in sys.stdin:
        line = line.strip()  # Remove espaços extras e quebras de linha
        if not line:
            continue
        
        parts = line.split(maxsplit=1)
        if not parts:
            continue

        try:
            d = int(parts[0])  # Lê a distância inicial
        except ValueError:
            continue

        evento = parts[1] if len(parts) > 1 else ""

        if evento.startswith("Consumo de combustível"):
            try:
                consumo = int(evento.split()[-1])
                if d == 0 and consumo == 0:
                    break
                gasto_combustivel.append((d, consumo / 100.0 + total_vazamentos))
            except ValueError:
                consumo = -1
        elif evento == "Vazamento":
            total_vazamentos += 1
            gasto_combustivel.append((d, consumo / 100.0 + total_vazamentos))
        elif evento == "Mecânico":
            total_vazamentos = 0
            gasto_combustivel.append((d, consumo / 100.0))
        elif evento == "Posto de gasolina":
            tamanho_tanque = 0
            gasto_combustivel.append((d, consumo / 100.0 + total_vazamentos))
            
            for i in range(len(gasto_combustivel) - 1):
                tamanho_tanque += gasto_combustivel[i][1] * (gasto_combustivel[i + 1][0] - gasto_combustivel[i][0])
            
            tamanho_min_tanque = max(tamanho_min_tanque, tamanho_tanque)
            gasto_combustivel.clear()
            gasto_combustivel.append((d, consumo / 100.0 + total_vazamentos))
        elif evento == "Destino":
            tamanho_tanque = 0
            gasto_combustivel.append((d, consumo / 100.0 + total_vazamentos))
            
            for i in range(len(gasto_combustivel) - 1):
                tamanho_tanque += gasto_combustivel[i][1] * (gasto_combustivel[i + 1][0] - gasto_combustivel[i][0])
            
            tamanho_min_tanque = max(tamanho_min_tanque, tamanho_tanque)
            print(f"{tamanho_min_tanque:.3f}")
            sys.stdout.flush()
            
            gasto_combustivel.clear()
            total_vazamentos = 0
            consumo = -1
            tamanho_min_tanque = 0

if __name__ == "__main__":
    main()

File: 6-DescidaDaSerra/main.cpp
/*
 * Nome: Rodrigo de Freitas Lima
 * NUSP: 12547510
*/

#include <iostream>
#include <sstream>
#include <string>
#include <algorithm>
#include <vector>
#include <iomanip>
#include <limits>

using namespace std;

int main() {
    string line; // Linha de entrada
    float tamanhoMinTanque = 0;
    vector<pair<float, float>> gastoCombustivel; // <distancia, consumo>
    int totalVazamentos = 0; // Contador de vazamentos
    int consumo = -1; // consumo atual

    while (getline(cin, line)) {
        // Remove carriage return if present (helps with file redirection)
        if (!line.empty() && line.back() == '\r') {
            line.pop_back();
        }

        istringstream iss(line);
        int d;

        // Le a distância inicial
        if (!(iss >> d)) {
            continue; // Skip invalid lines
        }

        string evento;
        // Le o restante da linha como uma string completa
        getline(iss, evento);
        
        // Remove espaços a esquerda
        evento.erase(0, evento.find_first_not_of(" "));
        // Remove espaços extras no final
        if (!evento.empty()) {
            evento.erase(evento.find_last_not_of(" ") + 1);
        }

        // Parsing de eventos usando string::find para maior robustez
        if (evento.find("Consumo de combustível") == 0) {
            // Extrair o número após "Consumo de combustível"
            size_t pos = evento.find_last_of(" ");
            if (pos != string::npos) {
                istringstream numStream(evento.substr(pos + 1));
                if (numStream >> consumo) {
                    if (d == 0 && consumo == 0) { // Encerra o programa se a distancia e o consumo forem 0
                        break;
                    }
                    pair<float, float> par = make_pair(d, consumo / 100.0 + totalVazamentos);
                    gastoCombustivel.push_back(par);
                } else {
                    consumo = -1; // Entrada invalida
                }
            }
        } else if (evento == "Vazamento") {
            totalVazamentos++;
            pair<float, float> par = make_pair(d, consumo / 100.0 + totalVazamentos);
            gastoCombustivel.push_back(par);
        } else if (evento == "Mecânico") {
            totalVazamentos = 0;
            pair<float, float> par = make_pair(d, consumo / 100.0);
            gastoCombustivel.push_back(par);
        } else if (evento == "Posto de gasolina") {
            float tamanhoTanque = 0;
            pair<float, float> par = make_pair(d, consumo / 100.0 + totalVazamentos);
            gastoCombustivel.push_back(par);

            for(int i = 0; i < gastoCombustivel.size() - 1; i++) {
                // tanque = somatorio[i = 0, n-1](consumo * distancia entre pontos)
                tamanhoTanque += gastoCombustivel[i].second * (gastoCombustivel[i + 1].first - gastoCombustivel[i].first);
            }

            tamanhoMinTanque = max(tamanhoMinTanque, tamanhoTanque);

            gastoCombustivel.clear();
            gastoCombustivel.push_back(par);
        } else if (evento == "Destino") {
            float tamanhoTanque = 0;
            pair<float, float> par = make_pair(d, consumo / 100.0 + totalVazamentos);
            gastoCombustivel.push_back(par);

            for(int i = 0; i < gastoCombustivel.size() - 1; i++) {
                // tanque = somatorio[i = 0, n-1](consumo * distancia entre pontos)
                tamanhoTanque += gastoCombustivel[i].second * (gastoCombustivel[i + 1].first - gastoCombustivel[i].first);
            }

            tamanhoMinTanque = max(tamanhoMinTanque, tamanhoTanque);

            // imprime ate 3 casas decimais
            cout << fixed << setprecision(3) << tamanhoMinTanque << endl;
            cout.flush(); // Ensure output is flushed

            gastoCombustivel.clear();
            totalVazamentos = 0;
            consumo = -1;
            tamanhoMinTanque = 0;
        }
    }

    return 0;
}